Mathlib/Algebra/Lie/Sl2.lean

```diff
@@ -0,0 +1,186 @@
+/-
+Copyright (c) 2024 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.Algebra.Lie.OfAssociative
+import Mathlib.LinearAlgebra.Eigenspace.Basic
+import Mathlib.LinearAlgebra.FreeModule.StrongRankCondition
+
+/-!
+
+# The Lie algebra `sl₂` and its representations
+
+The Lie algebra `sl₂` is the unique simple Lie algebra of minimal rank, 1, and as such occupies a
+distinguished position in the general theory. This file provides some basic definitions and results
+about `sl₂`.
+
+## Main definitions:
+ * `Sl2Triple`: a structure representing a triple of elements in a Lie algebra which satisfy the
+   standard relations for `sl₂`.
+ * `Sl2Triple.HasPrimitiveVectorWith`: a structure representing a primitive vector in a
+   representation of a Lie algebra relative to a distinguished `sl₂` triple.
+ * `Sl2Triple.HasPrimitiveVectorWith.exists_nat`: the eigenvalue of a primitive vector must be a
+   natural number if the representation is finite-dimensional.
+
+-/
+
+-- TODO Find home for this, maybe generalise and put in `Mathlib.Order.SuccPred.Basic`?
+lemma Nat.exists_not_and_succ_of_exists_and_not_zero {p : ℕ → Prop} (H : ∃ n, p n) (H' : ¬ p 0) :
+    ∃ n, ¬ p n ∧ p (n + 1) := by
+  classical
+  let k := Nat.find H
+  have hk : p k := Nat.find_spec H
+  suffices 0 < k from
+    ⟨k - 1, Nat.find_min H <| Nat.pred_lt this.ne', by rwa [Nat.sub_add_cancel this]⟩
+  by_contra! contra
+  rw [nonpos_iff_eq_zero.mp contra] at hk
+  exact H' hk
+
+variable (R L M : Type*) [CommRing R] [LieRing L] [LieAlgebra R L]
+  [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+
+open LieModule
+
+/-- An `sl₂` triple within a Lie ring `L` is a triple of elements `h`, `e`, `f` obeying relations
+which ensure that the Lie subalgebra they generate is either equivalent to `sl₂` or trivial.
+
+Note that since `h = ⁅e, f⁆`, we do not include it as part of the data. -/
+@[ext] structure Sl2Triple where
```

Comment:
Have you also considered a predicate that takes `e`, `f` and `h` as arguments?

Reply:
I did consider this and I don't have a good answer: it seems like either design would be fine. I'm happy to switch or even to push up a commit demonstrating the change if you'd like to see it.

Reply:
I think we will rarely (never) want to consider the type of all sl2-triples in a Lie algera. Hence I'm leaning slightly towards the predicate version. If you don't mind testing it out, please go ahead!

Reply:
Nice! Do you think this PR is also a good place to add an example of the "canonical" sl2-triple? Namely the one for sl2 itself?

Reply:
This raises issues about what should import what which I'd rather defer.